### PR TITLE
Add farm management support

### PR DESCRIPTION
**Pull request by chl178**
## Background:
Our previous discussions have highlighted the need to provide more flexibility and efficiency in managing wiki farms in Canasta, which is a Docker-based MediaWiki distribution. The initial implementation included the ability to run multiple wikis (or a wiki farm) within a single container, which was a significant stride towards our project's goal. However, for even more user-friendly and effective management of wiki farms, the enhancement of Canasta's CLI became evident.
## Summary
This PR introduces two new commands to the CLI for managing wiki farms: add and remove. Furthermore, we've updated the create command to directly create a wiki farm in a Canasta instance using a wikis.yaml file.

Importantly, our design philosophy in adding wiki farm support ensures the continuity of the original functionality. If you choose not to use the wiki farm related features, all other command operations will work as before. This makes the adoption of wiki farm features completely optional, allowing you to continue using the CLI in the way you're familiar with if desired.
## Table of Contents
- Enhancements
  - create
  - extension
  - skin
- New Commands
  - add
  - remove
- Docker-Compose Structure
- Next Steps
- Version Migration

### create
**Description:** Creates a Canasta installation. Enhanced to support wiki farm setup with the `-f` flag.

**Usage:**
`sudo go run canasta.go create [flags]`
- `-p, --path`: Canasta directory.
- `-o, --orchestrator`: Orchestrator to use for installation (default: "docker-compose").
- `-i, --id`: Canasta instance ID.
- `-w, --wiki`: Name of the wiki.
- `-n, --domain-name`: Domain name (default: "localhost").
- `-a, --WikiSysop`: Initial wiki admin username.
- `-s, --password`: Initial wiki admin password.
- `-f, --yamlfile`: Initial wiki yaml file for wiki farm setup.
- `-k, --keep-config`: Keep the config files on installation failure.

**YAML Format for Wiki Farm:**
To create a wiki farm, you need to provide a YAML file with the following format:
```yaml
wikis:
  - id: [WIKI_ID] # Example: "wiki1"
    url: [WIKI_URL] # Example: "mywiki1.example.com"
    name: [WIKI_SITENAME] # Example: "mywiki1"
```
`sudo go run canasta.go create -f [yamlfile] # Example: "wikis.yaml`

### extension
**Description:** Manage Canasta extensions. Enhanced to target a specific wiki within the farm using the `-w` flag.

**Subcommands:**
- `list`: Lists all the installed Canasta extensions.
- `enable`: Enables specified extensions.
- `disable`: Disables specified extensions.

**Usage:**
`sudo go run canasta.go extension [subcommand] [flags]`
**Flags:**
- `-i, --id`: Specifies the Canasta instance ID.
- `-p, --path`: Specifies the Canasta installation directory.
- `-w, --wiki`: Specifies the ID of a specific wiki within the Canasta farm.
- `-v, --verbose`: Enables verbose output.

### add
**Description:** Adds a new wiki to a Canasta instance.

**Usage:**
`sudo go run canasta.go add [flags]`
**Flags:**
- `-w, --wiki`: ID of the new wiki.
- `-u, --url`: URL of the new wiki.
- `-s, --site-name`: Name of the new wiki site.
- `-p, --path`: Path to the new wiki.
- `-i, --id`: Canasta instance ID.
- `-o, --orchestrator`: Orchestrator to use for installation (default: "docker-compose").
- `-d, --database`: Path to the existing database dump.

### remove
**Description:** Removes a wiki from a Canasta instance.

**Usage:**
`sudo go run canasta.go remove [flags]`
**Flags:**
- `-w, --wiki`: ID of the wiki to be removed.
- `-p, --path`: Path to the wiki.
- `-i, --id`: Canasta instance ID.

## Docker-Compose Structure
Our current docker compose structure is as follows:
```
config/
	CommonSettings.php
	Caddyfile
	CanastaFooterIcon.php
	composer.local.json
	default.vcl
        wikis.yaml
  	settings/
                *.php
  	wiki1/
		Settings.php
    		*.php
  	wiki2/
    		Settings.php
    		*.php
images/
        wiki1/
  	wiki2/
```
Each wiki's settings file is contained within a directory of their name under the config directory. These settings include a CommonSettings.php file.
Eache wiki's images are stored in their corresponding file under images/[WIKIID].

In the new version, we consider all Canasta instances as wiki farms, even if there is only one wiki in that instance.Therefore, manual imports of settings and images must be placed in the corresponding wiki ID's folder.

## Version Migration
To make it more convenient for users to migrate to the new version, the new Canasta CLI will automatically detect when used, and then create configuration files such as YAML, images folder, etc. If the number of wikis in the farm is more than 1, then the LocalSettings.php will be renamed to CommonSettings.php.

## Next Steps:

we will initiate server tests to evaluate the practicality and performance of these updates. Your support and collaboration in this testing phase will be crucial to ensure that our modifications not only align with technical specifications but also cater to real-world usage scenarios.

### CODE REVIEW

**Comment by chl178** [comment]
### For people who want to test it locally
As this is an early version for wiki farm support for Canasta, it currently does not support CLI installation and can only be manually installed using docker-compose. You may refer to the official docker-compose manual installation [documentation](https://github.com/CanastaWiki/Canasta-Documentation/blob/main/setup.md#manual-installation) for guidance.
The [test template](https://github.com/CanastaWiki/Canasta-DockerCompose/pull/45) for docker-compose.

1. Modify your `hosts` file to add an alias for 127.0.0.1 and restart the Apache service.
Screenshot from 2023-06-04 19-54-27
2. Download my canasta code by `git clone git@github.com:chl178/Canasta.git -b farm` 
3. Create an image by `docker build -t canastafarm . `
4. Download the CLI and use sudo go run canasta.go [commands] to use it

**Comment by chl178** [comment]
## Discussion: Program Structure
I would like to initiate a discussion regarding two key elements of our current implementation. This involves the naming convention of our farm management commands and the structure of our docker-compose. Your feedback is crucial in guiding the development of these features.

### 1. Farm Management Command Structure
Currently, we have farm management commands add and remove which share a resemblance to our existing create and delete commands. To distinguish between these functionalities, we're considering the following options:

Place add and remove under a parent farm command: This would result in farm add and farm remove commands. The benefit of this approach is it groups farm-related operations under a single command, improving command organization.

Maintain add and remove at the same level as other commands: This approach keeps the add and remove commands at the same level as other commands, which is how we're currently operating.

Please share your thoughts on these two options or suggest alternate approaches.

### 2. Docker-Compose Structure
Our current docker-compose structure is as follows:

arduino
Copy code
config/
  CommonSettings.php
  Setting1.php
  Setting2.php
  wiki1/
    LocalSettings.php
    settings/
  wiki2/
    LocalSettings.php
    settings/
Each wiki's settings file is contained within a directory of their name under the config directory. These settings include a CommonSettings.php file.

We're trying to decide the naming convention for this common settings file. Should we keep it as CommonSettings.php or change it to LocalSettings.php? Please weigh in with your opinions and any alternate suggestions.

Looking forward to your valuable input on these matters.

**Comment by olsonjaredm** [comment]
I like using add/remove for the commands (I can't come up with any better terms). 

As far as PHP file structure I'd say default to CommonSettings.php for the upper level with the expectation that for single-wiki 'farms' this file is a stub and ignored by admins.
Then each of the subfolders for individual wikis contain their own LocalSettings.php. I also like the idea of it concatenating all PHP files in the folders so admins can rename them what they like.

File: cmd/add/add.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	err = mediawiki.InstallOne(instance.Path, name, domain, wikipath, instance.Orchestrator)
+	if err != nil {
+		return err
+	}
+	err = restart.Restart(instance)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Successfully Added the Wiki '" + name + "in Canasta instance '" + instance.Id + "'...")
```
Comment (jeffw16) [inline cmd/add/add.go:133]: shouldn't there be a space between `name +"` and `in Canasta`?

File: cmd/add/add.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	addCmd.Flags().StringVarP(&siteName, "site-name", "s", "", "Name of the new wiki site")
+	addCmd.Flags().StringVarP(&instance.Path, "path", "p", pwd, "Path to the new wiki")
+	addCmd.Flags().StringVarP(&instance.Id, "id", "i", "", "Canasta instance ID")
+	addCmd.Flags().StringVarP(&instance.Orchestrator, "orchestrator", "o", "docker-compose", "Orchestrator to use for installation")
+	addCmd.Flags().StringVarP(&databasePath, "database", "d", "", "Path to the existing database dump")
+	return addCmd
+}
+
+// addWiki accepts the Canasta instance ID, the name, domain and path of the new wiki, and the initial admin info, then creates a new wiki in the instance.
+func AddWiki(name, domain, wikipath, siteName, databasePath string, instance config.Installation) error {
```
Comment (jeffw16) [inline cmd/add/add.go]: Is it possible for any of these params to be non-null but empty? If so, how is that handled?

File: cmd/remove/remove.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	addCmd.Flags().StringVarP(&wikiName, "wiki", "w", "", "ID of the wiki")
+	addCmd.Flags().StringVarP(&instance.Path, "path", "p", pwd, "Path to the new wiki")
+	addCmd.Flags().StringVarP(&instance.Id, "id", "i", "", "Canasta instance ID")
+	return addCmd
+}
+
+// addWiki accepts the Canasta instance ID, the name, domain and path of the new wiki, and the initial admin info, then creates a new wiki in the instance.
+func RemoveWiki(name string, instance config.Installation) error {
```
Comment (jeffw16) [inline cmd/remove/remove.go:51]: What happens if this is removing the last wiki in the Canasta instance?

File: internal/canasta/canasta.go [hunk 4]
```diff
@@ -27,13 +28,15 @@ func CloneStackRepo(orchestrator, canastaId string, path *string) error {
 	*path += "/" + canastaId
 	logging.Print(fmt.Sprintf("Cloning the %s stack repo to %s \n", orchestrator, *path))
 	repo := orchestrators.GetRepoLink(orchestrator)
-	err := git.Clone(repo, *path)
+	err := git.Cloneb(repo, *path, "CLI")
```
Comment (jeffw16) [inline internal/canasta/canasta.go:31]: What is `git.Cloneb`? Is it a typo or is it different than `git.Clone`?

File: internal/canasta/canasta.go [hunk 5]
```diff
@@ -44,12 +47,180 @@ func CopyEnv(envPath, domainName, path, pwd string) error {
 	if err != nil {
 		return fmt.Errorf(output)
 	}
-	if err := SaveEnvVariable(path+"/.env", "MW_SITE_SERVER", "https://"+domainName); err != nil {
+	_, domainNames, _, err := farmsettings.ReadWikisYaml(yamlPath)
+	if err != nil {
+		return err
+	}
+	if err := SaveEnvVariable(path+"/.env", "MW_SITE_SERVER", "http://"+domainNames[0]); err != nil {
```
Comment (jeffw16) [inline internal/canasta/canasta.go]: Let's make sure this isn't `http://`. There should be an option to have `https://` at the very least

File: internal/canasta/canasta.go [hunk 6]
```diff
@@ -44,12 +47,180 @@ func CopyEnv(envPath, domainName, path, pwd string) error {
...
+	// Copy SettingsTemplate.php
+	err, output := execute.Run("", "cp", path+"/config/SettingsTemplate.php", dirPath+"/Settings.php")
+	if err != nil {
+		return fmt.Errorf(output)
+	}
+	return nil
+}
+
+func RewriteSettings(path string, WikiNames []string) error {
+	for _, name := range WikiNames {
```
Comment (jeffw16) [inline internal/canasta/canasta.go:109]: Is `name` always guaranteed to be nonempty?

**Review by jeffw16 — CHANGES_REQUESTED** [review]

**Pull request merged**